scapy/contrib/pnio.py

```diff
@@ -41,37 +47,293 @@
     0xFF41:"PTCP-DelayResPDU-followup",
     0xFF42:"PTCP-DelayFuResPDU",
     0xFF43:"PTCP-DelayResPDU",
-    }
-for i in xrange(0x0100, 0x1000):
-    PNIO_FRAME_IDS[i] = "RT_CLASS_3"
-for i in xrange(0x8000, 0xC000):
-    PNIO_FRAME_IDS[i] = "RT_CLASS_1"
-for i in xrange(0xC000, 0xFC00):
-    PNIO_FRAME_IDS[i] = "RT_CLASS_UDP"
-for i in xrange(0xFF80, 0xFF90):
-    PNIO_FRAME_IDS[i] = "FragmentationFrameID"
+}
+
+
+def i2s_frameid(x):
+    if x in PNIO_FRAME_IDS:
+        return PNIO_FRAME_IDS[x]
+    elif 0x0100 <= x < 0x1000:
+        return "RT_CLASS_3 (%4x)" % x
+    elif 0x8000 <= x < 0xC000:
+        return "RT_CLASS_1 (%4x)" % x
+    elif 0xC000 <= x < 0xFC00:
+        return "RT_CLASS_UDP (%4x)" % x
+    elif 0xFF80 <= x < 0xFF90:
+        return "FragmentationFrameID (%4x)" % x
+    return x
+
+
+def s2i_frameid(x):
+    try:
+        idx = PNIO_FRAME_IDS.values().index(x)
+        return PNIO_FRAME_IDS.keys()[idx]
+    except ValueError:
+        pass
+    if x == "RT_CLASS_3":
+        return 0x0100
+    elif x == "RT_CLASS_1":
+        return 0x8000
+    elif x == "RT_CLASS_UDP":
+        return 0xC000
+    elif x == "FragmentationFrameID":
+        return 0xFF80
+    return x
 
 #################
 ## PROFINET IO ##
 #################
 
-class ProfinetIO(Packet):
+class ProfinetIO(scapy.packet.Packet):
     """Basic PROFINET IO dispatcher"""
-    fields_desc = [XShortEnumField("frameID", 0, PNIO_FRAME_IDS)]
-    overload_fields = {
-        Ether: {"type": 0x8892},
-        UDP: {"dport": 0x8892},
-        }
+    fields_desc = [
+        scapy.fields.XShortEnumField("frameID", 0, (i2s_frameid, s2i_frameid))
+    ]
 
     def guess_payload_class(self, payload):
         # For frameID in the RT_CLASS_* range, use the RTC packet as payload
-        if (self.frameID >= 0x0100 and self.frameID < 0x1000) or \
-                (self.frameID >= 0x8000 and self.frameID < 0xFC00):
-            from scapy.contrib.pnio_rtc import PNIORealTime
-            return PNIORealTime
+        if (0x0100 <= self.frameID < 0x1000) or (0x8000 <= self.frameID < 0xFC00):
+            return PNIORealTimeCyclicPDU
+        return super(ProfinetIO, self).guess_payload_class(payload)
+
+scapy.packet.bind_layers(scapy.layers.l2.Ether, ProfinetIO, type=0x8892)
```

Comment:
Thank you, @gpotter2, for your comment. I know just what you mean. Unfortunately, this is the only way to write robust Python, because Python import system is completely broken. Allow me to convince you with the following example:

test.py
```
import test2
def toto():
  print 'toto'
  test2.tutu()
def tata():
  print 'tata'
if '__main__' == __name__:
  toto()
```
test2.py
```
import test
def tutu():
  print 'tutu'
  test.tata()
```
This works:
```
$ python test.py
toto
tutu
tata
```
Now, let's write:
test.py
```
from test2 import tutu
def toto():
  print 'toto'
  tutu()
def tata():
  print 'tata'
if '__main__' == __name__:
  toto()
```
test2.py
```
from test import tata
def tutu():
  print 'tutu'
  tata()
```
This will break horribly:
```
$ python test.py
Traceback (most recent call last):
  File "test.py", line 1, in <module>
    from test2 import tutu
  File "/tmp/another/test2.py", line 1, in <module>
    from test import tata
  File "/tmp/another/test.py", line 1, in <module>
    from test2 import tutu
ImportError: cannot import name tutu
```

Unless Python import system is someday fixed, using import and specifying the full path at each invocation is the only way to prevent some import loops. Please bare in mind that using `import scapy.layers.l2 as l2` is also broken.

Now, my pool request does not need it. This is just defensive programming, in case one day it is needed.